source/auth/tests/mongodb-oidc.rst

```diff
@@ -4,146 +4,430 @@ MongoDB OIDC
-#. Perform a ``find`` operation.
-#. Clear the cache.
+Reauthentication
+================
 
-#. Set the ``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable to the location
-of the ``test_user1`` generated token file.
-#. Create a client with a url of the form ``mongodb://localhost:27018/?authMechanism=MONGODB-OIDC&authMechanismProperties=PROVIDER_NAME:aws&directConnection=true&readPreference=secondaryPreferred``.
-#. Perform a ``find`` operation.
+The driver MUST test reauthentication with MONGODB-OIDC for a read
+operation.
 
-#. Set the ``AWS_WEB_IDENTITY_TOKEN_FILE`` environment variable to the location
-of the ``test_user2`` generated token file.
-#. Create a client with a url of the form ``mongodb://localhost:27018/?authMechanism=MONGODB-OIDC&authMechanismProperties=PROVIDER_NAME:aws&directConnection=true&readPreference=secondaryPreferred``.
-#. Perform a ``find`` operation.
+Succeeds
+~~~~~~~~
+- Clear the cache.
+- Create request and refresh callbacks that return valid credentials
+  that will not expire soon.
+- Create a client with the callbacks and an event listener.  The following
+  assumes that the driver does not emit ``saslStart`` or ``saslContinue``
+  events.  If the driver does emit those events, ignore/filter them for the
+  purposes of this test.
+- Perform a ``find`` operation that succeeds.
+- Assert that the refresh callback has not been called.
+- Clear the listener state if possible.
+- Force a reauthenication using a ``failCommand`` of the form:
 
-#. Create a client with a url of the form  ``mongodb://localhost:27018/?authMechanism=MONGODB-OIDC&directConnection=true&readPreference=secondaryPreferred`` and the OIDC request callback.
-#. Assert that a ``find`` operation fails.
+.. code:: javascript
 
+    {
+      "configureFailPoint": "failCommand",
+      "mode": {
+        "times": 1
+      },
+      "data": {
+        "failCommands": [
+          "find"
+        ],
+        "errorCode": 391
+      }
+    }
 
-Invalid Callbacks
-=================
+.. note::
 
-Ensure the of the following conditions raise an error:
+  the driver MUST either use a unique ``appName`` or explicitly
+  remove the ``failCommand`` after the test to prevent leakage.
 
-#. Any callback returns null
-#. Any callback returns unexpected result
+- Perform another find operation that succeeds.
+- Assert that the refresh callback has been called once, if possible.
+- Assert that the ordering of list started events is [``find``],
+  , ``find``.  Note that if the listener stat could not be cleared then there
+  will and be extra ``find`` command.
+- Assert that the list of command succeeded events is [``find``].
+- Assert that a ``find`` operation failed once during the command execution.
+- Close the client.
 
-Cached Credentials
-==================
+Retries and Succeeds with Cache
```

Comment:
What is the purpose of this reauth test and the following one? (They appear to be doing the same thing but one has a success assertion and the other has a fail assertion). And the spec states that reauthentication must not be using cache values so I'm wondering if these 2 tests are actually needed?

Reply:
These two are covering the case where the server info has changed, but we are using stale server info.  I agree "Separate Connections Avoid Extra Callback Calls" can be merged with "Succeeds", will do.

Reply:
No actually, that would complicate the "Succeeds" test.  This test is about ensuring that we do not call the callbacks more than is necessary, by adding guards/checks on whether a cache value has been updated since we last used it.